### DISCUSSION TITLE
Timestream datasource backend evaluation

### DISCUSSION THREAD

**Discussion by sunker** [grafana/grafana#40744] [Data Sources]
Timestream is similar to Redshift and Athena in the sense that is's an AWS service and that the query language used is SQL. It's also different from them since the underlying data store is a time series database and not a relational/columnar database. 

The timesteam data source is a backend plugin built by Ryan from the edge team. Currently, it's using the timestream http data api (through the the AWS SDK for go) to query data. Sessions are provided by the grafana-aws-sdk.  

The redshift and athena data source both use the same backend architecture. The backend delegates the gRPC server api to SQLDS and redshift and athena has their own custom built go-sql driver and a custom rest api. SQLDS uses the sqlutil package from grafana-plugin-sdk-go to handle the conversion from sql rows to dataframes. Consequently, the backend for Redshift and Athena is simple, just containing a few hundred lines of code where the main part is about mapping api response types to gosql types. There are many reasons for us to reuse this architecure for all our sql-like data sources. It will save us time on maintenance, faster knowledge sharing, for our users it provides recognition of features between data sources etc etc.

We should therefore investigate if it's possible to rewrite the timestream backend to use SQLDS without introducing breaking changes.

**Comment by ryantxu** [comment]
For sure this is worth exploring, but I am a bit skeptical that the timeseries type maps to a SQL query result cleanly.  But there may be a multi-table or nested table SQL format I don't know about

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). Due to the age and number of responses to this discussion, we are deciding to close it. If this is something you would like to see in Grafana, feel free to open an [issue](https://github.com/grafana/grafana/issues) so the discussion can continue. Thank you!
